### PR TITLE
refactor(update-agent): get rid of gpt fork and TakePartition

### PR DESCRIPTION
**Pull request by TheButlah**
rebased on #459

### CODE REVIEW

File: io-utils/src/clamped.rs [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+        );
+        self.cursor = from_self_start;
+
+        Ok(self.cursor)
+    }
+}
+
+impl<RS: Read + Seek> Read for ClampedSeek<RS> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        if self.cursor >= self.clamp.end {
```
Comment (alekseifedotov) [inline io-utils/src/clamped.rs:109]: I don't think this short-circuiting is needed, as it will be handled by a regular code path.

**Review by alekseifedotov — APPROVED** [review]
I'm glad that you removed the gpt for.
I am astounded by the amount of boilerplate to make ClampedSeek.

**Comment by TheButlah** [comment]
marking as draft while I rework a bit more and test

**Comment by TheButlah** [comment]
> I am astounded by the amount of boilerplate to make ClampedSeek.

Yeah, its about 128 lines for the actual struct implementation. However it could be half if I didn't check for overflows etc.

Most of the boilerplate actually is in the tests, not sure how to improve that. I'm also not happy with how verbose the tests are. I think it may be a skill issue on my end, or maybe seeking and reading is just super edge case prone.

File: io-utils/src/clamped.rs (hunk 1 above)
Comment (TheButlah) [reply io-utils/src/clamped.rs]: it actually is necesssary otherwise we do an extraneous self.inner.read and/or panic on checked_sub

**Comment by TheButlah** [comment]
I've run two OTAs now on my pearl orb via --update-location and everything seems to have functioned correctly.
